services/yellow/src/api/Handlers/ChargeForOrderFullAmountHandler.cs

```diff
@@ -1,52 +1,60 @@
+        
+        public ChargeForOrderFullAmountHandler(IOrderService orderService, IPaymentOrderAdapter paymentOrderAdapter)
         {
             _orderService = orderService;
-            _creditCardPaymentService = creditCardPaymentService;
+            _paymentOrderAdapter = paymentOrderAdapter;
         }
         
-        protected async override Task<OrderFullAmountCharged> Handle(ChargeForOrderFullAmountRequest message)
+        public async Task Handle(ChargeForOrderFullAmountRequest message, IMessageHandlerContext context)
         {
-            var order = await Ensure(message, r => _orderService.FindOneById(r.OrderId));
-
-            var amountToPay = order.Price.Amount;
+            try
+            {
+                var validateResult = ValidateRequest(message);
+                if (!validateResult.IsValid())
+                    throw new ValidationException(validateResult);
             
-            var chargeResult = (message.AlternativePaymentMethod is not null)
-                ? await ChargeOrder(order, amountToPay, message.AlternativePaymentMethod)
-                : await ChargeOrder(order, amountToPay);
+                var order = HandlerHelper.Ensure(await _orderService.FindOneById(message.OrderId))!;
+
+                var amountToPay = order.Price.Amount;
             
-            if (message.AlternativePaymentMethod is not null)
-                await AddPaymentMethodToOrder(order, message.AlternativePaymentMethod.CreditCard, chargeResult.AuthorizePaymentToken);
+                var chargeResult = (message.AlternativePaymentMethod is not null)
+                    ? await _paymentOrderAdapter.ChargeOrder(order, amountToPay, message.AlternativePaymentMethod)
+                    : await _paymentOrderAdapter.ChargeOrder(order, amountToPay);
 
-            // TODO : Should go in a separate handler (and should retrieve it from the used payment token)
-            var lastPaymentMethod = order.PaymentMethods!.Last();
-            await _orderService.AddTransaction(order, chargeResult.Transaction.GetOrderTransaction(lastPaymentMethod));
+                if (message.AlternativePaymentMethod is not null)
+                    await _paymentOrderAdapter.AddPaymentMethodToOrder(order, message.AlternativePaymentMethod.CreditCard, chargeResult.AuthorizePaymentToken);
+
+                var lastPaymentMethod = order.PaymentMethods!.Last();
+                await _orderService.AddTransaction(order, chargeResult.Transaction.GetOrderTransaction(lastPaymentMethod));
             
-            return new OrderFullAmountCharged(message.OrderId);
-        }
-        
-        protected override object Fail(ChargeForOrderFullAmountRequest requestMessage)
-        {
-            return new ChargeForOrderFullAmountFailed(requestMessage.OrderId);
+                await context.Publish(new OrderFullAmountCharged(message.OrderId));
+                await context.Reply(new Response<bool>(true));

```

Comment:
This `Response` possibly breaks the purple saga because it's not expecting a response.
`Response<T>` should be returned only to the ambassador's calls, as a quick-feedback.